htdocs/js/tree/notebook_tree_model.js

```diff
@@ -0,0 +1,1691 @@
+                
+                var history_len = 0;
+                if(that.histories_[node.gistname]) {
+                    history_len = that.histories_[node.gistname].length;
+                }
+
+                that.on_show_history.notify({
+                    node: node,
+                    history_len: history_len
+                });
+
+                node.is_open = true;
+            });
+    },
+
+    load_everything: function() {
+    
+        var that = this,
+            opts;
+        return Promise.all([
+            rcloud.get_users(),
+            that.get_starred_info(),
+            that.get_recent_info(),
+            rcloud.get_gist_sources(),
+            rcloud.config.get_user_option(['notebook-path-tips', 'tree_sort_order', 'tree_filter_date'])
+        ]).spread(function(all_the_users, starred_info, recent_info, gist_sources, user_options) {
+            opts = user_options;
+            that.path_tips_ = user_options['notebook-path-tips'];
+            
+            that.gist_sources_ = gist_sources;
+            _.extend(that.notebook_info_, starred_info.notebooks);
+            for(var r in recent_info.notebooks) {
+                // add a special flag for notebooks that we only know about from recent list
+                // we won't show them in the tree until they're opened
+                if(!that.notebook_info_[r]) {
+                    that.notebook_info_[r] = recent_info.notebooks[r];
+                    that.notebook_info_[r].recent_only = true;
+                }
+            }
+            _.extend(that.num_stars_, recent_info.num_stars); // (not currently needed)
+            all_the_users = _.union(all_the_users, _.compact(_.pluck(starred_info.notebooks, 'username')));
+
+            var featured_tree;
+
+            return Promise.all([rcloud.config.get_current_notebook(),
+                                that.get_featured()
+                                ])
+                .spread(function(current, featured_notebooks) {
+                    that.current_ = current;
+                    that.num_stars_ = starred_info.num_stars;
+                    featured_tree = featured_notebooks;
+
+                })
+                .then(function() {
+
+                    var alls_root = that.populate_users(all_the_users);
+
+                    return [
+                        featured_tree,
+                        that.populate_interests(starred_info.notebooks),
+                        that.populate_friends(all_the_users),
+                        alls_root
+                    ].filter(function(t) { return !!t; });
+
+                });
+        })
+        .then(function(data) {
+
+            // initial assignment: 
+            this.tree_data_ = data;
+
+            this.update_filter({
+                prop: 'tree_filter_date',
+                value: opts['tree_filter_date']
+            });
+
+            this.update_sort_type(opts['tree_sort_order'], true);   
```

Comment:
remember that these `opts` will be undefined the first time the code is run - probably worth testing with all options set to `null`...